meteor/server/api/triggeredActions.ts

```diff
@@ -0,0 +1,183 @@
+import { Meteor } from 'meteor/meteor'
+import { check, Match } from '../../lib/check'
+import { registerClassToMeteorMethods } from '../methods'
+import { literal, getRandomId, protectString, makePromise, unprotectString } from '../../lib/lib'
+import { ServerResponse, IncomingMessage } from 'http'
+import { logger } from '../logging'
+import { ShowStyleBases, ShowStyleBaseId } from '../../lib/collections/ShowStyleBases'
+import { MethodContext, MethodContextAPI } from '../../lib/api/methods'
+import { ShowStyleContentWriteAccess } from '../security/showStyle'
+import { PickerPOST, PickerGET } from './http'
+import {
+	DBTriggeredActions,
+	TriggeredActionId,
+	TriggeredActions,
+	TriggeredActionsObj,
+} from '../../lib/collections/TriggeredActions'
+import { NewTriggeredActionsAPI, TriggeredActionsAPIMethods } from '../../lib/api/triggeredActions'
+import { SystemWriteAccess } from '../security/system'
+
+export function createTriggeredActions(
+	showStyleBaseId: ShowStyleBaseId | null,
+	base?: Partial<Pick<DBTriggeredActions, '_rank' | 'triggers' | 'actions' | 'name'>>
+) {
+	const id: TriggeredActionId = getRandomId()
+	TriggeredActions.insert(
+		literal<TriggeredActionsObj>({
+			_id: id,
+			_rank: base?._rank ?? 0,
+			name: base?.name,
+			showStyleBaseId,
+			_rundownVersionHash: '',
+			actions: base?.actions ?? [],
+			triggers: base?.triggers ?? [],
+		})
+	)
+	return id
+}
+
+export function removeTriggeredActions(triggeredActionId: TriggeredActionId) {
+	TriggeredActions.remove(triggeredActionId)
+}
+
+PickerPOST.route('/actionTriggers/upload/:showStyleBaseId?', (params, req: IncomingMessage, res: ServerResponse) => {
+	res.setHeader('Content-Type', 'text/plain')
+
+	const showStyleBaseId: ShowStyleBaseId | undefined = protectString(params.showStyleBaseId)
+
+	check(showStyleBaseId, Match.Maybe(String))
+	let content = ''
+
+	const replace: boolean = params.query === 'replace'
+
+	try {
+		if (showStyleBaseId) {
+			const showStyleBase = ShowStyleBases.findOne(showStyleBaseId)
+			if (!showStyleBase) {
+				throw new Meteor.Error(
+					404,
+					`Restore Action Triggers: ShowStyle "${showStyleBaseId}" could not be found`
+				)
+			}
+		}
+
+		const body = req.body
+		if (!body) throw new Meteor.Error(400, 'Restore Action Triggers: Missing request body')
+
+		if (typeof body !== 'string' || body.length < 10)
+			throw new Meteor.Error(400, 'Restore Action Triggers: Invalid request body')
+
+		const triggeredActions = JSON.parse(body) as DBTriggeredActions[]
+		check(triggeredActions, Array)
+
+		// set new showStyleBaseId
+		for (let i = 0; i < triggeredActions.length; i++) {
+			check(triggeredActions[i]._id, String)
+			check(triggeredActions[i].name, Match.Optional(String))
+			check(triggeredActions[i].triggers, Array)
+			check(triggeredActions[i].actions, Array)
+			triggeredActions[i].showStyleBaseId = showStyleBaseId ?? null
```

Comment:
Just a thought, would it be good to do a specific check for empty-string?
Since it it is falsy, the same way null is, but wont be picked-up when querying mongo for null.

Reply:
I'm adding an undefined check at line in 54, so that any string (as opposed to any falsy value) is treated as a ShowStyleId, so an empty string should cause it to try and find a matching ShowStyleBase and fail with message.